lib/pure/asyncdispatch.nim

```diff
@@ -1851,6 +1909,105 @@ proc send*(socket: AsyncFD, data: string,
+    ev.vd = nil
+    ev.cb = nil
+    ev.vFD = InvalidVirtualFD
+
+  if vd.virtualHandles.len == 0:
+    # lazy close the physical event with the Dispatcher
+    # The main reason we need this is to make
+    # hasPendingOperations still work
+    # without modifying the ioselector code.
+    vd.close
+
+proc initVirtualEventDispatcher(p: PDispatcher) =
+  p.vd.virtualMuxHandle = newAsyncEvent()
+
+  proc virtualMuxCB(fd: AsyncFD): bool {.closure,gcsafe.} =
+    # find the virtual events that triggered the physical event and
+    # add them to the callback list.
+    # Not very efficient, but requires the least coordination between threads.
+    
+    proc vcbFactory(vfd: VirtualFD) : proc() {.closure, gcsafe.} =
+      let ev = p.vd.virtualHandles[vfd]
+      result = proc() {.closure,gcSafe.} =
+        if ev.cb(vfd.AsyncFD):
+          # the convention is that if the callback returns true, we unregister the event.
+          ev.unregister
+
+    for vfd, ev in p.vd.virtualHandles:
+      withLockIfThreads ev.eventLock:
+        if ev.triggered:
+          ev.triggered = false
+          p.callbacks.addLast(vcbFactory(vfd))
+
+    # always return false b/c we never want to unregister this event
+    return false
+
+  p.vd.virtualMuxHandle.addEvent(virtualMuxCB)
+
+proc newVirtualAsyncEvent*(): VirtualAsyncEvent =
+  ## Creates a new thread-safe ``VirtualAsyncEvent`` object.
+  ##
+  ## New ``VirtualAsyncEvent`` object is not automatically registered with
+  ## dispatcher like ``AsyncSocket``.
+  result = cast[VirtualAsyncEvent](allocShared0(sizeof(VirtualAsyncEventImpl)))
+  result.cb = nil
+  result.vFD = InvalidVirtualFD
+  when compileOption("threads"):
+    initRLock result.eventLock
+
+proc trigger*(ev: VirtualAsyncEvent) =
+  ## Sets new ``VirtualAsyncEvent`` to signaled state.
+  withLockIfThreads ev.eventLock:
+    if ev.vFD == InvalidVirtualFD:
+      # triggering on an event that is not registered is a noop.
+      return
+    ev.triggered = true
+
+    # send the signal to wake up the dispatcher thread.
+    trigger(ev.vd.virtualMuxHandle)
+
+proc addEvent*(ev: VirtualAsyncEvent, cb: Callback) =
+  ## Registers callback ``cb`` to be called when ``ev`` will be signaled
+  doAssert(ev.vFD == InvalidVirtualFD, "Event is already registered in the queue!")
+
+  let p = getGlobalDispatcher()
+
+  if p.vd.virtualMuxHandle == nil:
+    # lazy register the physical event with the Dispatcher
+    # The main reason we need this is to make hasPendingOperations still work
+    # without modifying the ioselector code.
+    initVirtualEventDispatcher(p)
+
+  withLockIfThreads ev.eventLock:
+    ev.cb = cb
+    ev.vFD = p.vd.nextVirtualHandle
+    p.vd.nextVirtualHandle = VirtualFD(int(p.vd.nextVirtualHandle) + 1)
+    p.vd.virtualHandles[ev.vFD] = ev
+    ev.vd = p.vd
+
+proc close*(ev: VirtualAsyncEvent) =
+  ## Closes event ``ev``.
+  doAssert(ev.vFD != InvalidVirtualFD, "Must unregister Event before you close it!")
```

Comment:
IIRC for pretty much all other things in asyncdispatch, the ``close`` proc will implicitly unregister the FD. Please keep that consistent.

Reply:
Actually, this is safer than what happens in other places in asyncdispatch.

If you look at `AsyncEvent`, the following happens:

It calls `close` on the underlying `ioselector`, which closes it at the kernel level, and dealloc's the memory, but it doesn't actually unregister the FD:

https://github.com/nim-lang/Nim/blob/233455a6856e33486c534efc0dd8bb3db1d1e601/lib/pure/asyncdispatch.nim#L1546
https://github.com/nim-lang/Nim/blob/233455a6856e33486c534efc0dd8bb3db1d1e601/lib/pure/ioselects/ioselectors_epoll.nim#L104

Maybe that is a bug?

Reply:
*sigh*

Yeah, it's not consistent with sockets at all then...

https://github.com/nim-lang/Nim/blob/5288b546dab6db1bb7edae64813111b83c0a2643/lib/pure/asyncdispatch.nim#L1320-L1334